src/Core/Currency/ExchangeRateProvider.php

```diff
@@ -0,0 +1,293 @@
+<?php
+/**
+ * 2007-2019 PrestaShop SA and Contributors
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.txt.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to http://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2019 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+namespace PrestaShop\PrestaShop\Core\Currency;
+
+use PrestaShop\CircuitBreaker\Contract\CircuitBreakerInterface;
+use PrestaShop\Decimal\Number;
+use PrestaShop\PrestaShop\Core\Currency\Exception\CurrencyFeedException;
+use PrestaShop\PrestaShop\Core\Foundation\Filesystem\FileSystem;
+use SimpleXMLElement;
+
+/**
+ * Retrieves the exchange rate of a currency (based on the default currency). It uses a circuit breaker
+ * to avoid being blocked in case of network problems and it saves each of its request in a cache to be
+ * able to have a fallback response.
+ */
+class ExchangeRateProvider
+{
+    /**
+     * This url was set in the _PS_CURRENCY_FEED_URL_ const but it is not accessible in every
+     * context because it is weirdly defined in defines_uri.inc.php So it is safer to define
+     * it properly here.
+     */
+    const CURRENCY_FEED_URL = 'http://api.prestashop.com/xml/currencies.xml';
+
+    const DEFAULT_EXCHANGE_RATE = 1.0;
+
+    const CLOSED_ALLOWED_FAILURES = 3;
+    const CLOSED_TIMEOUT_SECONDS = 1;
+
+    const OPEN_ALLOWED_FAILURES = 3;
+    const OPEN_TIMEOUT_SECONDS = 2;
+    const OPEN_THRESHOLD_SECONDS = 3600; // 1 hour
+
+    /** @var string */
+    private $currencyFeedUrl;
+
+    /** @var string */
+    private $defaultCurrencyIsoCode;
+
+    /** @var CircuitBreakerInterface */
+    private $remoteServiceProvider;
+
+    /** @var string */
+    private $cacheDir;
+
+    /** @var string */
+    private $cacheFile;
+
+    /** @var string */
+    private $sourceIsoCode;
+
+    /** @var array */
+    private $currencies;
```

Comment:
```suggestion
    private $currencies = [];
```

Reply:
Because https://github.com/PrestaShop/PrestaShop/pull/14972/files#diff-52a57e335fe7b5677855d9b6a76ae835R122 can fail if it's not an array

Reply:
and because of https://github.com/PrestaShop/PrestaShop/pull/14972/files#diff-52a57e335fe7b5677855d9b6a76ae835R233